unity-renderer/Assets/DCLServices/WearablesCatalogService/WebInterfaceWearablesCatalogService.cs

```diff
@@ -0,0 +1,374 @@
+            CheckForRequestsTimeOutsAsync(serviceCts.Token).Forget();
+            CheckForRequestsByContextTimeOutsAsync(serviceCts.Token).Forget();
+
+            // Check unused wearables (to be removed from our catalog) only every [TIME_TO_CHECK_FOR_UNUSED_WEARABLES] seconds
+            CheckForUnusedWearablesAsync(serviceCts.Token).Forget();
+        }
+
+        public void Initialize(
+            WearablesWebInterfaceBridge wearablesWebInterfaceBridge,
+            BaseDictionary<string, WearableItem> wearablesCatalog)
+        {
+            webInterfaceBridge = wearablesWebInterfaceBridge;
+            WearablesCatalog = wearablesCatalog;
+            Initialize();
+        }
+
+        public void Dispose()
+        {
+            Destroy(this);
+        }
+
+        public async UniTask<WearableItem[]> RequestOwnedWearablesAsync(string userId, int pageNumber, int pageSize, CancellationToken ct) =>
+            await RequestWearablesByContextAsync(userId, null, null, $"{OWNED_WEARABLES_CONTEXT}{userId}", false, ct);
+
+        public async UniTask<WearableItem[]> RequestBaseWearablesAsync(CancellationToken ct) =>
+            await RequestWearablesByContextAsync(null, null, new [] { BASE_WEARABLES_COLLECTION_ID }, BASE_WEARABLES_CONTEXT, false, ct);
+
+        public async UniTask<WearableItem[]> RequestThirdPartyWearablesByCollectionAsync(string userId, string collectionId, CancellationToken ct) =>
+            await RequestWearablesByContextAsync(null, null, null, $"{THIRD_PARTY_WEARABLES_CONTEXT}_{collectionId}", true, ct);
+
+        public async UniTask<WearableItem[]> RequestWearablesAsync(string[] wearableIds, CancellationToken ct) =>
+            await RequestWearablesByContextAsync(null, wearableIds, null, string.Join(",", wearableIds.ToArray()), false, ct);
+
+        public async UniTask<WearableItem> RequestWearableAsync(string wearableId, CancellationToken ct)
+        {
+            if (WearablesCatalog.TryGetValue(wearableId, out WearableItem wearable))
+            {
+                if (wearablesInUseCounters.ContainsKey(wearableId))
+                    wearablesInUseCounters[wearableId]++;
+
+                return wearable;
+            }
+
+            ct.ThrowIfCancellationRequested();
+
+            UniTaskCompletionSource<WearableItem> taskResult;
+
+            if (!awaitingWearableTasks.ContainsKey(wearableId))
+            {
+                taskResult = new UniTaskCompletionSource<WearableItem>();
+                awaitingWearableTasks.Add(wearableId, taskResult);
+
+                // We accumulate all the requests during the same frames interval to send them all together
+                pendingWearablesToRequest.Add(wearableId);
+            }
+            else
+                taskResult = awaitingWearableTasks[wearableId];
+
+            ct.RegisterWithoutCaptureExecutionContext(() => taskResult.TrySetCanceled());
+            return await taskResult.Task;
```

Comment:
cancellation is handled in the line above: `ct.RegisterWithoutCaptureExecutionContext(() => taskResult.TrySetCanceled());`. Should it be different?

Reply:
I didn't notice that:
```
ct.RegisterWithoutCaptureExecutionContext(() => taskResult.TrySetCanceled());
```
It's incorrect as we reuse `taskResult` from the cache. Thus, we are attaching several independent `CancellationToken`'s to the same `CompletionSource`.
Instead, we should be attaching it to the task that projects onto `CompletionSource` (`taskResult.Task` returns a fresh copy every time is called).
But I would redesign the flow of waiting for `CompletionSources` as I proposed in one of the previous threads: it would eliminate this creation of `CompletionSource` and the `async` flow would look more natural/sequential